### ISSUE TITLE
Add ability to safely release unintentionally dragged handle

### ISSUE THREAD

**Issue by DIV-on-github** [Slicer/Slicer#6038]
## Use cases
Adapted from https://discourse.slicer.org/t/rfe-add-ability-to-safely-disengage-unintentionally-dragged-handle/20484

# Unintentionally grabbing a handle (instead of rotating, etc.)

In the 3D viewport a common operation would be to use the mouse to rotate by left-clicking (or, strictly speaking, left-button-pressing) & dragging. The same sequence is also used for other tasks, though, such as resizing a ROI, or moving a fiducial marker (from the Markups module).

It happens occasionally that the user intends to rotate the 3D view, but ends up inadvertently grabbing the handle of a ROI or a fiducial marker. 
In such situations, currently — even when the user realises mid-action that the wrong task is underway — there is no alternative but to complete the action, meaning that the ROI or fiducial marker is undesirably moved (with no recourse to restore it to precisely its original condition).

# Unintentionally _not_ grabbing a handle (but rather rotating, etc.)
On some other occasions a user might have want to resize the ROI, but ends up accidentally rotating the 3D view. Similarly, it could happen to a user attempting to drag a fiducial marker too. 
In such situations, currently — even when the user realises mid-action that the wrong task is underway — there is no alternative but to complete the action, meaning that the view is undesirably rotated (with no recourse to restore it to precisely its original condition).

## A solution

In such cases it'd be helpful to be able to safely disengage _immediately_, without completing the action. In other software a similar functionality can be accomplished by pressing the `Esc` key on the keyboard before the mouse button is released.

## Alternatives

The other option, of course, would be a global Undo functionality, but it lacks immediacy, and also it is questionable whether actions such as rotating a view should be recorded in the undo 'stack'.  

There are also _some_ ways of avoiding the problem:  lock fiducial markers (they're unlocked by default);  hide the ROI (it needs to be displayed in order to set its extents, and while setting the extents it may be desirable to rotate the view);  and save a Scene View (users would not like to have to create dozens of Scene Views 'just in case').  However they rely on the user to anticipate the problem and then perform additional tasks, which is both a less robust and more onerous workflow.

**Comment by jamesobutler** [comment]
xref #4576

**Comment by lassoan** [comment]
> In such cases it'd be helpful to be able to safely disengage immediately, without completing the action. In other software a similar functionality can be accomplished by pressing the Esc key on the keyboard before the mouse button is released.

This is not an easy option to implement with the current design, as the node is immediately changed once interaction starts. If we disengage at any time it is already too late because the node has already changed.

We could save the state of the markup node. Simply saving and restoring a control point position and state may not be sufficient, because the markup node (or any modules that react to markup node changes) may decide to perform any additional internal state changes in response to interactions.

We already have undo/redo infrastructure for this, it is just not tested and optimized enough. The first step should be fix and enable undo/redo infrastructure for markups. Then Escape could be abort the operation and call Undo. Undo/redo would also address many other issues and enhancements that users have been asking for.

**Comment by pieper** [comment]
> Simply saving and restoring a control point position and state may not be sufficient, because the markup node (or any modules that react to markup node changes) may decide to perform any additional internal state changes in response to interactions.

I know we've been over this before, but I'll say again that if code is listening and responding to mouse movements then it should just as easily be able to respond to a movement that puts the point back where it was at the start of the action if escape is triggered - no harm in that, it's just one more mouse move.

**Comment by lassoan** [comment]
When you move the point then not just the position is changed but for example the point normal, the node where that point was picked on, etc. is also be saved. Custom markups may store even more information. If we restore the point position on Escape but none of the additionally collected information that could lead to errors.

**Comment by pieper** [comment]
Then any code that relies on such extra information should have an 'OnMoveCanceled' method or similar that resets any of those assumptions.  Or we should finish the undo mechanism and the cancel operation is in effect, 'complete the move and then undo'.

I'm just saying that I agree with @DIV-on-github that the current behavior is non-standard and not user friendly.  The expected behavior is not hard to describe, even if it does add some work on the implementation.

**Comment by lassoan** [comment]
> the current behavior is non-standard and not user friendly

I agree that hitting Escape while dragging to cancel an operation can be considered standard, so it would be nice to provide it in Slicer.

> Or we should finish the undo mechanism and the cancel operation is in effect, 'complete the move and then undo'.

Yes, exactly.

We must implement undo for many other reasons, for example, if you realize that you unintentionally moved a point after you release the mouse button then Escape key will not save you. For this reason, instead of introducing an "undo light" mechanism just for managing markup point movements (`OnMoveCanceled`), I would rather spend that time with testing, fixing, and re-enabling the proper undo mechanism.

**Comment by pieper** [comment]
Fair enough.  I would really like to see undo/redo finally make it into the application as a whole, or at least the most critical portions.

**Comment by lassoan** [comment]
There are just a few specific issues that would need to be fixed before we enable undo/redo for markups. Probably not a lot of work, but it would be probably better to work on it after Slicer5 is released.
